Assignment of parameters in pulse `Schedule`/`ScheduleBlock` doable through parameter name

arthurostrauss opened the pull request:
<!--
⚠️ If you do not respect this template, your pull request will be closed.
⚠️ Your pull request title should be short detailed and understandable for all.
⚠️ Also, please add a release note file using reno if the change needs to be
  documented in the release notes.
⚠️ If your pull request fixes an open issue, please link to the issue.

- [ ] I have added the tests to cover my changes.
- [ ] I have updated the documentation accordingly.
- [ ] I have read the CONTRIBUTING document.
-->

### Summary

This PR extends the previous usage of the method `assign_parameters` of `Schedule`and `ScheduleBlock` by enabling the specification of the parameter name instead of the parameter directly. This assignment by name also works when assigning values to a `ParameterVector`(the length of the list of parameter values should in this case match the length of the vector).

### Details and comments

arthurostrauss committed:
Added possibility of assigning Parameter and ParameterVector by name

It is now possible to specify in the mapping the names of the parameters instead of the parameters themselves to assign parameters to pulse `Schedule`s and `ScheduleBlock`s. It is even possible to assign all the parameters of a ParameterVector by just specifying its name and setting as a value a list of parameter values.

wshanks commented:
@arthurostrauss Is there previous discussion of this? It is convenient, but it is not something that `QuantumCircuit` allows either. I wonder if we want to discuss both contexts and keep behavior consistent. There is also a tradeoff between convenience and the potential for assigning to more parameters than intended (the user can avoid the ambiguity by using parameters but we should consider how much room for ambiguity to allow). We have struggled with string names versus parameters in the Qiskit Experiments calibration module in the past (with blocks like [this one](https://github.com/Qiskit-Extensions/qiskit-experiments/blob/cf6222de654d227f9aac20e3b13aa7d87f8c36c3/qiskit_experiments/calibration_management/calibrations.py#L1049-L1052)).

arthurostrauss commented:
The motivation behind such feature lies in the will to integrate in a closer form the primitives V2 to pulse level control, as the former do carry `BindingArrays` storing the parameter names only. The idea would be to assign parameters to the schedule based solely on the information of the BindingArrays. For instance, one could imagine in the future when Qiskit Pulse will be fully JAX compatible, doing a subroutine using the primitives (with `DynamicsBackend`) and simulating the pipeline at the pulse level. Instead of binding the parameters before Jit compiling the workflow, one could accelerate the pipeline by compiling over the parametrized schedule, and bind the values using vectorized modelling of Jax, significantly improving performance. I am already working myself on such use case, happy to tell you more if relevant.

arthurostrauss commented:
By the way after checking, it is possible to use `QuantumCircuit.assign_parameters()`with strings representing the parameter name. Cf this note in the documentation under the method:
 .. note::
            When ``parameters`` is given as a mapping, it is permissible to have keys that are
            strings of the parameter names; these will be looked up using :meth:`get_parameter`.
            You can also have keys that are :class:`.ParameterVector` instances, and in this case,
            the dictionary value should be a sequence of values of the same length as the vector.

            If you use either of these cases, you must leave the setting ``flat_input=False``;
            changing this to ``True`` enables the fast path, where all keys must be
            :class:`.Parameter` instances.

wshanks commented:
@arthurostrauss That sounds nice! My main concern was adding something to pulse `assign_parameters` that was not in circuit `assign_parameters` but I had missed that circuits supported string parameters. I think this should be fine conceptually then (I haven't looked at it closely enough to approve right now).

arthurostrauss commented:
Hi @wshanks, any updates on this?

wshanks commented:
@arthurostrauss For my part, I had an initial concern about deviating from `QuantumCircuit.assign_parameters` that you addressed, and then I just have not had time to look at this more closely. I will when I have time. I don't want to block this though if @nkanazawa1989 or @TsafrirA are able to look at it before me.

wshanks commented on qiskit/pulse/parameter_manager.py:
```diff
@@ -415,7 +416,7 @@ def _unroll_param_dict(
                     raise PulseError(
                         f"Parameter vector '{parameter.name}' has length {len(parameter)},"
-                        f" but was assigned to a single value."
+                        f" but was assigned to {value}."
```
I wonder if the second conditional should print `value`. Maybe it would be hard to tell what the problem was if the length was long and just off by one element.

arthurostrauss replied:
Overall, the Pulse model of assigning parameters was very poor in checking the variables input (it was left to the class `ParameterExpression`during the binding. We could keep going this track and make the type checking, as well as the input type as minimal as possible (hence removing those warnings), to facilitate a further Jax compatible workflow (where such checks would have to be disabled to integrate the parameter assigmnent within a Jit compiled function. Another alternative for this last aspect is to add a class attribute disable_validation (similarly to what was proposed in #11029) to the `Parameter` class directly (to avoid any additional Python check on traced variables).

wshanks replied:
That is a reasonable path. My comment here was just about whether `value` should be printed if the lengths don't match below as well. Maybe we can just leave it printing the lengths for now.

wshanks commented on qiskit/pulse/parameter_manager.py:
```diff
@@ -424,7 +425,24 @@ def _unroll_param_dict(
                 out.update(zip(parameter, value))
             elif isinstance(parameter, str):
-                out[self.get_parameters(parameter)] = value
+                for param in self.parameters:
```
Oh, although the type annotation did not advertise it, `assign_parameters` already supported `str`. It is just the `ParameterVector` support that you added recently that did not support it. So we should think of this as a follow up to #12045.

arthurostrauss replied:
Assignment by string was in fact not supported before. It did not raise an error, but it didn't do proper parameter assignment.

arthurostrauss replied:
I realised that if you try the following code with the current version in the main (that I had implemented) you have an error, yielding the need to really support the string assignment:


`a = Parameter("a")
with pulse.build(name="test") as sched:
    pulse.play(pulse.Gaussian(100, a, 50), pulse.DriveChannel(0))`

Moreover, if you remove the part in the `_unroll_dict`method the special case (elif) for string handling (leaving the default case), you end up with the parameter not being assigned at all.

wshanks replied:
Hmm, that code doesn't error for me on `main`. Am I missing something from your example?

```python
a = Parameter("a")
with pulse.build(name="test") as sched:
    pulse.play(pulse.Gaussian(100, a, 50), pulse.DriveChannel(0))
```

However,

```python
sched.assign_parameters({"a": 1}).blocks
```

prints

    (Play(Gaussian(duration=100, sigma=50, amp=a, angle=0.0), DriveChannel(0)),)

So you are right about string assignment not working now. My previous test did not work properly because I tested `{a: 1}` and then `{"a": 1}` and did not account for the fact that the default is `inplace=True`.

arthurostrauss replied:
The code should have raised an error for the following reason: 
On the main, when you currently assign by the parameter by its name, the line 427 of the parameter_manager.py is failing: `out[self.get_parameters(parameter)] = value`.
In fact, the error is due to the fact that get_parameters() returns a list of parameters and not a parameter instance directly, making it an unusable key for the dict out. This is why we need to correct this with this PR.
Make sure you don't have the similar case as you depicted where the parameter has already been assigned (returning an empty list) for your test.

wshanks replied:
Ah, okay. I was just a little confused by your example above. I think you mean to include `assign_parameters({"a": 1})` in it? I do get the error with string assignment on `main` now. I think my `main` was old before.

wshanks commented on qiskit/pulse/parameter_manager.py:
```diff
@@ -424,7 +425,24 @@ def _unroll_param_dict(
+                        if not isinstance(value, Sequence):
+                            raise PulseError(
+                                f"ParameterVector '{param.vector.name}' has length {len(param.vector)},"
+                                f" but was assigned to a single value."
```
If we keep this, this should print the value like the case you modified above.

wshanks commented on releasenotes/notes/parameter_assignment_by_name_for_pulse_schedules-3a27bbbbf235fb9e.yaml:
```diff
@@ -0,0 +1,10 @@
+---
+prelude: >
+
+features_pulse:
```
I think we don't need a new release note since assignment with a string was already supported, just not for `ParameterVector`. The release note from #12045 could be modified, but I don't think it has to be.

wshanks replied:
Resolving since string assignment actually did not work before this.

wshanks reviewed (commented):
Although the type annotations do not advertise it, `assign_parameters` already supports `str`. What you are adding here is support for assigning a sequence to elements of a `ParameterVector` using the name of the `ParameterVector` like:

```python
from qiskit import pulse
from qiskit.circuit import ParameterVector

vec = ParameterVector("test", 2)

with pulse.builder.build() as sched:
    pulse.builder.play(pulse.Gaussian(100, vec[0], vec[1]), pulse.DriveChannel(0))

sched.assign_parameters({"theta": [1, 2]})
```

Since `ParameterVector` support was only added recently in #12045, this PR should be viewed as a follow up to it.

Going back to my original concern about deviation from the behavior of `QuantumCircuit`, I note that `QuantumCircuit.assign_parameters` does not allow assigning to a parameter vector by name. We should see if there is interest in such support for that case as well.

wshanks commented on test/python/pulse/test_parameter_manager.py:
```diff
@@ -515,6 +515,42 @@ def test_parametric_pulses_with_parameter_vector(self):
+        block += pulse.Play(waveform2, pulse.DriveChannel(10))
+        block += pulse.ShiftPhase(param_vec[0], pulse.DriveChannel(10))
+        block += pulse.ShiftPhase(param_vec[1], pulse.DriveChannel(10))
+        block.assign_parameters({"amp": 0.2, "sigma": 4, "param_vec": [3.14, 1.57]}, inplace=True)
```
Fun fact: you can also do `"param_vec[0]": 3.14, "param_vec[1]": 1.57`

wshanks commented on qiskit/pulse/parameter_manager.py:
```diff
@@ -424,7 +425,24 @@ def _unroll_param_dict(
+                    if param.name == parameter:
+                        out[param] = value
+                    elif (
+                        isinstance(param, ParameterVectorElement) and param.vector.name == parameter
```
I find this part to be getting a little complicated. Could we lift the string handling to before this loop? So in a preceding loop, loop over `parameter_binds` and make a new copy that has all `Parameter` and `ParameterVector` keys by replacing any strings with the appropriately named matches?

wshanks reviewed (commented):
Sorry for the slightly redundant comments in the previous review. I was shifting comments around and hadn't meant to submit the review yet. Here were a couple other comments.

wshanks commented on releasenotes/notes/parameter_assignment_by_name_for_pulse_schedules-3a27bbbbf235fb9e.yaml:
```diff
@@ -0,0 +1,10 @@
+---
+prelude: >
+
```
I think it is better not to include these lines.

```suggestion
```

wshanks reviewed (commented):
@arthurostrauss Could you address my comment about lifting up the string handling to earlier in the function and my minor suggestion about the release note? Otherwise, I think this looks good.

arthurostrauss committed:
Update parameter_assignment_by_name_for_pulse_schedules-3a27bbbbf235fb9e.yaml

Co-authored-by: Will Shanks <wshaos@posteo.net>

arthurostrauss committed:
Reshaped string parameter assignment

Corrected mistake in string assignment

Corrected mistake in string assignment

wshanks commented on qiskit/pulse/parameter_manager.py:
```diff
@@ -410,21 +411,23 @@ def _unroll_param_dict(
             A dictionary from parameter to value.
         """
         out = {}
+        param_name_dict = {param.name: param for param in self.parameters}
```
This does not account for the fact that more than one parameter could have the same name, which pulse allows but circuit does not (it gives an error when trying to add the second parameter).

I think would be best to loop over `self.parameters` and make one dict shaped like `dict[str, list[Parameter | ParameterVector]]` mapping each name to the parameters with that name. Then make a new dict and loop over `parameter_binds` copying in the key and values for when the key is a `Parameter` or `ParameterVector` and replacing the key when it is a string with an entry for each value in the first dict for that string (could be more than one parameter). Then loop over that dict like below but without accounting for string keys.

You may want to account for a name being used for a parameter and a parameter vector. There is validation for vectors but not for scalars. If you tried to assign to a list a string name used by both a `ParameterVector` and a `Parameter`, it would not catch the assignment of a list to a `Parameter` here.

Maybe it is worth testing duplicated names?

arthurostrauss replied:
So just to clarify, the expected behaviour is that every `Parameter` carrying the same name should be assigned to the same value? Or should it be that a list of values could be passed such that all parameters carrying the same name are taken into account? If it's the latter case, then how can we resolve this when using string parameter assignment?

wshanks replied:
Every `Parameter` carrying the same name should be assigned the same value when a string is assigned a value. Also, every `ParameterVector` should be assigned the same list of values.

wshanks commented on releasenotes/notes/parameter_assignment_by_name_for_pulse_schedules-3a27bbbbf235fb9e.yaml:
```diff
@@ -0,0 +1,8 @@
+---
+features_pulse:
+  - |
+    It is now possible to assign parameters to pulse `Schedule`and `ScheduleBlock` objects by specifying
```
Sphinx syntax:

```suggestion
    It is now possible to assign parameters to pulse :class:`.Schedule`and :class:`.ScheduleBlock` objects by specifying
```

arthurostrauss committed:
Update releasenotes/notes/parameter_assignment_by_name_for_pulse_schedules-3a27bbbbf235fb9e.yaml

Co-authored-by: Will Shanks <wshaos@posteo.net>

arthurostrauss committed:
Enabled string assignment for multiple params carrying same name

The check is now based on the value type to infer if assignment should be done on Parameters or ParameterVectors

Removed unnecessary import from utils

Corrected string assignment

Correction part 2

Corrected test

The inplace=True argument was preventing the reuse of a parametrized waveform in the schedule, making the test fail

arthurostrauss committed:
Reformat

wshanks reviewed (approved):
This looks good to me now!
